src/common/host_device_vector.cu

```diff
@@ -358,19 +358,33 @@ struct HostDeviceVectorImpl {
     return data_h_;
   }
 
-  void Reshard(const GPUDistribution& distribution) {
+  void Shard(const GPUDistribution& distribution) {
     if (distribution_ == distribution) { return; }
-    CHECK(distribution_.IsEmpty() || distribution.IsEmpty());
-    if (distribution.IsEmpty()) {
-      LazySyncHost(GPUAccess::kWrite);
-    }
+    CHECK(distribution_.IsEmpty());
+    CHECK(!distribution.IsEmpty());
     distribution_ = distribution;
     InitShards();
   }
 
-  void Reshard(GPUSet new_devices) {
+  void Shard(GPUSet new_devices) {
     if (distribution_.Devices() == new_devices) { return; }
-    Reshard(GPUDistribution::Block(new_devices));
+    Shard(GPUDistribution::Block(new_devices));
+  }
+
+  void Unshard() {
+    CHECK(!distribution_.IsEmpty());
+    LazySyncHost(GPUAccess::kWrite);
+    distribution_ = GPUDistribution::Block(GPUSet::Empty());
+    InitShards();
+  }
+
+  void Reshard(const GPUDistribution& distribution) {
```

Comment:
This actually calls `InitShards()` twice. Just syncing to the host and updating/reinitializing the shards should be enough.

Reply:
Done.